packages/go_router/example/lib/shell_route_top_route.dart

```diff
@@ -0,0 +1,326 @@
+              ),
+            ],
+          ),
+        ],
+      ),
+    ],
+  );
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      routerConfig: _router,
+    );
+  }
+}
+
+/// Builds the "shell" for the app by building a Scaffold with a
+/// BottomNavigationBar, where [child] is placed in the body of the Scaffold.
+class ScaffoldWithNavBar extends StatelessWidget {
+  /// Constructs an [ScaffoldWithNavBar].
+  const ScaffoldWithNavBar({
+    required this.title,
+    this.routeName,
+    required this.child,
+    super.key,
+  });
+
+  /// The title to display in the AppBar.
+  final String title;
+
+  /// The name of the current route.
+  final String? routeName;
+
+  /// The widget to display in the body of the Scaffold.
+  /// In this sample, it is a Navigator.
+  final Widget child;
+
+  @override
+  Widget build(BuildContext context) {
+    final PreferredSizeWidget leadingButton = _buildAppBar(context);
+    return Scaffold(
+      body: child,
+      appBar: leadingButton,
+      bottomNavigationBar: BottomNavigationBar(
+        items: const <BottomNavigationBarItem>[
+          BottomNavigationBarItem(
+            icon: Icon(Icons.home),
+            label: 'A Screen',
+          ),
+          BottomNavigationBarItem(
+            icon: Icon(Icons.business),
+            label: 'B Screen',
+          ),
+          BottomNavigationBarItem(
+            icon: Icon(Icons.notification_important_rounded),
+            label: 'C Screen',
+          ),
+        ],
+        currentIndex: _calculateSelectedIndex(context),
+        onTap: (int idx) => _onItemTapped(idx, context),
+      ),
+    );
+  }
+
+  /// Animates the AppBar title and leading button when it changes.
+  PreferredSizeWidget _buildAppBar(BuildContext context) {
+    return PreferredSize(
+      preferredSize: const Size.fromHeight(kToolbarHeight),
+      child: AnimatedSwitcher(
+        duration: const Duration(milliseconds: 120),
+        child: AppBar(
+          key: routeName == null ? UniqueKey() : ValueKey<String>(routeName!),
+          title: Text(title),
+          leading: _buildLeadingButton(context),
+        ),
+      ),
+    );
+  }
+
+  /// Builds the app bar leading button using the current location Uri
+  Widget _buildLeadingButton(BuildContext context) {
+    final RouteMatchList currentConfiguration =
+        GoRouter.of(context).routerDelegate.currentConfiguration;
+    final RouteMatch lastMatch = currentConfiguration.last;
+    final Uri location = lastMatch is ImperativeRouteMatch
+        ? lastMatch.matches.uri
+        : currentConfiguration.uri;
+    final bool canPop = location.pathSegments.length > 1;
+    final Widget button = canPop
+        ? BackButton(onPressed: GoRouter.of(context).pop)
+        : const SizedBox.shrink(key: ValueKey<String>('NoLeadingButton'));
+    return button;
+  }
```

Comment:
I had to manually create the leading button since this is in a scaffold above the children. Is there a better way of doing this? For context, auto_route has `AutoLeadingButton`